### PR TITLE
Agenda and My Schedule

### PR DESCRIPTION
**Pull request by LuigiR0jas**
## Description

Please include a summary of the change. Also, include any additional information that you think is important for reviewers to know. Link to a related issue if applicable.

## How has this been tested?

Please mark the tests that you ran to verify your changes. If difficult to test, consider providing instructions so reviewers can test.

- [ ] Manual testing
- [ ] System tests
- [ ] Unit tests
- [ ] None

## Checklist

- [ ] CI pipeline is passing
- [ ] My code follows the conventions of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented on my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (if applicable)
- [ ] I have added seed data to the database (if applicable)

## Release tasks

Add any tasks that need to be done before/after the release of this feature.

## Screenshots/Loom

This section is relevant in case we want to share progress with the team, otherwise, it can be omitted.

### CODE REVIEW

File: app/views/sessions/_card.html.erb [hunk 1]
```diff
@@ -1,35 +1,62 @@
-<div class="w-full p-5 bg-white rounded-lg">
-    <div class="flex flex-row items-center mb-6">
-      <% if speaker.profile.image.attached? %>
-        <%= image_tag url_for(profile.image), class: "text-red w-12 h-12 rounded-full mr-4" %>
-      <% else %>
-        <div class="flex flex-row items-center justify-center w-12 h-12 p-2 mr-4 rounded-full min-w-12 min-h-12 bg-slate-300">
-            <%= image_tag url_for("user_avatar.svg") %>
+<div class="flex flex-col w-full p-5 mb-4 bg-white rounded-[20px]">
```
Comment (Sergio-e) [inline app/views/sessions/_card.html.erb:1]: I think we should start using [strict_locals](https://api.rubyonrails.org/classes/ActionView/Template.html#method-i-strict_locals-21) when using partials, as it makes them easier to reuse.

File: app/models/conference.rb [hunk 2]
```diff
@@ -16,4 +16,10 @@ class Conference < ApplicationRecord
   def self.ransackable_attributes(_auth_object = nil)
     %w[name]
   end
+
+  def session_dates
```
Comment (Sergio-e) [inline app/models/conference.rb:20]: I suggest removing the cache. The query seems to be very fast


image

File: app/models/session.rb [hunk 3]
```diff
@@ -32,7 +32,18 @@ class Session < ApplicationRecord
...
+
+  after_commit :invalidate_cache
+
   def self.ransackable_attributes(_auth_object = nil)
     %w[title]
   end
+
+  private
+
+  def invalidate_cache
```
Comment (Sergio-e) [inline app/models/session.rb:46]: Reminder - remove

File: app/views/sessions/_card.html.erb [hunk 4]
```diff
@@ -1,35 +1,62 @@
...
-    </div>
-    <h1 class="mb-4 text-xl italic font-bold text-red"><%= session.title %></h1>
-    <div class="flex flex-row flex-wrap items-center gap-2 mb-[18px]">
-    <% session.tags.each do |tag| %>
-      <div class="bg-white border border-blue-2 rounded-lg px-4 py-[6px] text-blue-2">
-      <%= tag.name %>
+
+      <div class="flex items-center px-1">
+        <%= button_to session_path(id: session.id, on_date: request.params[:on_date]), method: :put, form_class: "flex items-center" do %>
+          <% if current_user.sessions.include?(session) %>
```
Comment (Sergio-e) [inline app/views/sessions/_card.html.erb:23]: I don't think we need to be worrying about n+1 given the expected DB size, however, if time allows, consider having a single query in sessions/index and then sending a local to the partial that determines if the icon needs to be displayed or not.

File: app/views/sessions/_card.html.erb [hunk 5]
```diff
@@ -1,35 +1,62 @@
...
+
+      <div class="flex items-center mb-3">
+        <%= inline_svg_tag("icons/clock.svg", class: "w-[16px] h-[16px] mr-1") %>
+        <span class="font-sans text-sm text-gray-600"><%= session.starts_at.strftime("%H:%M") %>
+          - <%= session.ends_at.strftime("%H:%M") %></span>
+      </div>
+
+      <div class="flex items-center">
+        <%= inline_svg_tag("icons/location.svg", class: "w-[16px] h-[16px] mr-1") %>
+        <span class="font-sans text-sm text-gray-600"><%= session.location.name %></span>
```
Comment (Sergio-e) [inline app/views/sessions/_card.html.erb:58]: Reminder - use `&`

File: app/controllers/application_controller.rb [hunk 6]
```diff
@@ -4,9 +4,12 @@ class ApplicationController < ActionController::Base
...
 
-  helper_method :current_profile
+  helper_method :current_profile, :current_conference
 
   private
 
   def current_profile = current_user&.profile
+
+  # TODO: Must change after implementing multi-conference support
+  def current_conference = Conference.last
```
Comment (Sergio-e) [inline app/controllers/application_controller.rb:14]: Let's filter sessions based on the current_conference

File: app/views/sessions/index.html.erb [hunk 7]
```diff
@@ -1,3 +1,28 @@
...
+        <% end %>
+      </div>
+
+      <div class="flex items-center justify-center ml-3 p-2 bg-white rounded-full min-w-[40px] min-h-[40px]">
+        <%= inline_svg_tag("icons/filters.svg", size: "18") %>
+      </div>
+    </div>
+
+    <div class="flex flex-col w-full h-full">
+      <% @sessions.each do |session| %>
```
Comment (andresag4) [inline app/views/sessions/index.html.erb:23]:
```
09:34:35 web.1  | [primary]   ↳ app/views/sessions/_card.html.erb:24
09:34:35 web.1  |   Rendered sessions/_card.html.erb (Duration: 9.8ms | Allocations: 3363)
09:34:35 web.1  | [primary]   Session Exists? (0.5ms)  SELECT 1 AS one FROM "sessions" INNER JOIN "sessions_users" ON "sessions"."id" = "sessions_users"."session_id" WHERE "sessions_users"."user_id" = ? AND "sessions"."id" = ? LIMIT ?  [["user_id", 2], ["id", 24], ["LIMIT", 1]]
09:34:35 web.1  | [primary]   ↳ app/views/sessions/_card.html.erb:24
09:34:35 web.1  |   Rendered sessions/_card.html.erb (Duration: 2.7ms | Allocations: 1610)
09:34:35 web.1  | [primary]   Session Exists? (0.7ms)  SELECT 1 AS one FROM "sessions" INNER JOIN "sessions_users" ON "sessions"."id" = "sessions_users"."session_id" WHERE "sessions_users"."user_id" = ? AND "sessions"."id" = ? LIMIT ?  [["user_id", 2], ["id", 25], ["LIMIT", 1]]
09:34:35 web.1  | [primary]   ↳ app/views/sessions/_card.html.erb:24
09:34:35 web.1  |   Rendered sessions/_card.html.erb (Duration: 3.2ms | Allocations: 1614)
09:34:35 web.1  | [primary]   Session Exists? (0.5ms)  SELECT 1 AS one FROM "sessions" INNER JOIN "sessions_users" ON "sessions"."id" = "sessions_users"."session_id" WHERE "sessions_users"."user_id" = ? AND "sessions"."id" = ? LIMIT ?  [["user_id", 2], ["id", 26], ["LIMIT", 1]]
09:34:35 web.1  | [primary]   ↳ app/views/sessions/_card.html.erb:24
09:34:35 web.1  |   Rendered sessions/_card.html.erb (Duration: 1.9ms | Allocations: 1611)
09:34:35 web.1  | [primary]   Session Exists? (0.4ms)  SELECT 1 AS one FROM "sessions" INNER JOIN "sessions_users" ON "sessions"."id" = "sessions_users"."session_id" WHERE "sessions_users"."user_id" = ? AND "sessions"."id" = ? LIMIT ?  [["user_id", 2], ["id", 27], ["LIMIT", 1]]
09:34:35 web.1  | [primary]   ↳ app/views/sessions/_card.html.erb:24
09:34:35 web.1  |   Rendered sessions/_card.html.erb (Duration: 1.8ms | Allocations: 1611)
09:34:35 web.1  | [primary]   Session Exists? (0.5ms)  SELECT 1 AS one FROM "sessions" INNER JOIN "sessions_users" ON "sessions"."id" = "sessions_users"."session_id" WHERE "sessions_users"."user_id" = ? AND "sessions"."id" = ? LIMIT ?  [["user_id", 2], ["id", 28], ["LIMIT", 1]]
09:34:35 web.1  | [primary]   ↳ app/views/sessions/_card.html.erb:24
09:34:35 web.1  |   Rendered sessions/_card.html.erb (Duration: 2.2ms | Allocations: 1610)
09:34:35 web.1  | [primary]   Session Exists? (0.6ms)  SELECT 1 AS one FROM "sessions" INNER JOIN "sessions_users" ON "sessions"."id" = "sessions_users"."session_id" WHERE "sessions_users"."user_id" = ? AND "sessions"."id" = ? LIMIT ?  [["user_id", 2], ["id", 29], ["LIMIT", 1]]
09:34:35 web.1  | [primary]   ↳ app/views/sessions/_card.html.erb:24
09:34:35 web.1  |   Rendered sessions/_card.html.erb (Duration: 2.9ms | Allocations: 1614)
09:34:35 web.1  | [primary]   Session Exists? (0.6ms)  SELECT 1 AS one FROM "sessions" INNER JOIN "sessions_users" ON "sessions"."id" = "sessions_users"."session_id" WHERE "sessions_users"."user_id" = ? AND "sessions"."id" = ? LIMIT ?  [["user_id", 2], ["id", 30], ["LIMIT", 1]]
09:34:35 web.1  | [primary]   ↳ app/views/sessions/_card.html.erb:24
09:34:35 web.1  |   Rendered sessions/_card.html.erb (Duration: 2.9ms | Allocations: 1613)
```


```
09:36:07 web.1  | [primary]   Session Exists? (0.4ms)  SELECT 1 AS one FROM "sessions" INNER JOIN "sessions_users" ON "sessions"."id" = "sessions_users"."session_id" WHERE "sessions_users"."user_id" = ? AND "sessions"."id" = ? LIMIT ?  [["user_id", 2], ["id", 30], ["LIMIT", 1]]
09:36:07 web.1  | [primary]   ↳ app/views/sessions/_card.html.erb:24
09:36:07 web.1  |   Rendered collection of sessions/_card.html.erb [30 times] (Duration: 209.0ms | Allocations: 121788)
```

File: app/views/sessions/index.html.erb [hunk 8]
```diff
@@ -1,3 +1,28 @@
...
+        <%= render partial:'sessions/card', locals: { session: session, speaker: session.speakers.first } %>
+      <% end %>
```
Comment (andresag4) [inline app/views/sessions/index.html.erb:25]:
```suggestion
      <%= render partial: 'sessions/card', collection: @sessions, as: :session %>
```

File: app/views/sessions/_card.html.erb [hunk 9]
```diff
@@ -1,35 +1,67 @@
...
       </div>
-    </div>
-    <h1 class="mb-4 text-xl italic font-bold text-red"><%= session.title %></h1>
-    <div class="flex flex-row flex-wrap items-center gap-2 mb-[18px]">
-    <% session.tags.each do |tag| %>
-      <div class="bg-white border border-blue-2 rounded-lg px-4 py-[6px] text-blue-2">
-      <%= tag.name %>
+
+      <div class="flex items-center px-1">
+        <% if session&.attendees&.include?(current_user) %>
```
Comment (andresag4) [inline app/views/sessions/_card.html.erb:25]:
```suggestion
        <% if session.attendees.include?(current_user) %>
```

File: app/controllers/attendees_controller.rb [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    end
+
+    redirect_back_or_to(sessions_path, params: params[:starts_at])
+  end
+
+  def destroy
+    if @session.attendees.delete(current_user)
+      flash[:notice] = I18n.t("controllers.attendees.remove_user.notice")
+    else
+      flash[:alert] = I18n.t("controllers.attendees.remove_user.alert")
```
Comment (andresag4) [inline app/controllers/attendees_controller.rb:18]: remove flash alerts

File: app/controllers/attendees_controller.rb [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    end
+
+    redirect_back_or_to(sessions_path, params: params[:starts_at])
+  end
+
+  private
+
+  def set_session
+    @session = current_conference.sessions.find(params[:session_id])
```
Comment (andresag4) [inline app/controllers/attendees_controller.rb:27]: memoize

File: app/controllers/sessions_controller.rb [hunk 12]
```diff
@@ -1,9 +1,17 @@
...
   def show
-    @session = Session.find(params[:id])
+    @session = sessions.find(params[:id])
     @speaker = @session.speakers.first
   end
+
+  private
+
+  def sessions
+    @sessions ||= current_conference&.sessions
```
Comment (Sergio-e) [inline app/controllers/sessions_controller.rb:15]: No need to memoize queries that aren't expensive, plus you are only calling this method once per action so the memoization isn't being used. 
```suggestion
    current_conference&.sessions
```

File: app/views/sessions/_card.html.erb [hunk 13]
```diff
@@ -1,35 +1,67 @@
-<div class="w-full p-5 bg-white rounded-lg">
-    <div class="flex flex-row items-center mb-6">
-      <% if speaker.profile.image.attached? %>
-        <%= image_tag url_for(profile.image), class: "text-red w-12 h-12 rounded-full mr-4" %>
-      <% else %>
-        <div class="flex flex-row items-center justify-center w-12 h-12 p-2 mr-4 rounded-full min-w-12 min-h-12 bg-slate-300">
-            <%= image_tag url_for("user_avatar.svg") %>
+<%# TODO: remove when we add support for multiple speakers %>
```
Comment (Sergio-e) [inline app/views/sessions/_card.html.erb:1]: I suggest adding a padding/margin between the last card and the bottom navbar
image

File: app/views/layouts/_flash_message.html.erb [hunk 14]
```diff
@@ -1,3 +1,5 @@
+<%# locals: (message: nil) %>
```
Comment (Sergio-e) [inline app/views/layouts/_flash_message.html.erb:1]: I think this partial should always receive a `message`, so I suggest removing `nil` and raising an error if this local isn't present
```suggestion
<%# locals: (message:) %>
```

File: app/views/notifications/_card.html.erb [hunk 15]
```diff
@@ -1,19 +1,21 @@
+<%# locals: (profile: nil, message: nil) %>
```
Comment (Sergio-e) [inline app/views/notifications/_card.html.erb:1]:
```suggestion
<%# locals: (profile:, message:) %>
```

File: app/views/sessions/_filters.html.erb [hunk 16]
```diff
@@ -0,0 +1,30 @@
+<%# locals: (resource: nil) %>
```
Comment (Sergio-e) [inline app/views/sessions/_filters.html.erb:1]:
```suggestion
<%# locals: (resource:) %>
```

File: app/views/sessions/_card.html.erb [hunk 17]
```diff
@@ -1,35 +1,67 @@
-<div class="w-full p-5 bg-white rounded-lg">
```
Comment (Sergio-e) [inline app/views/sessions/_card.html.erb:1]: Let's add `session` as a required local

File: app/views/sessions/_card.html.erb [hunk 18]
```diff
@@ -1,35 +1,67 @@
...
-            <%= image_tag url_for("user_avatar.svg") %>
+<%# TODO: remove when we add support for multiple speakers %>
+<% speaker = session.speakers.first %>
+
+<div class="flex flex-col w-full p-5 mb-4 bg-white rounded-[20px]">
+  <div class="flex flex-col w-full">
+    <div class="flex justify-between w-full mb-4">
+      <div class="flex">
+        <% if speaker&.image&.attached? %>
+          <%= image_tag url_for(speaker&.image), class: "text-red w-12 h-12 rounded-full mr-4" %>
```
Comment (Sergio-e) [inline app/views/sessions/_card.html.erb:9]:
```suggestion
          <%= image_tag url_for(speaker.image), class: "text-red w-12 h-12 rounded-full mr-4" %>
```

File: app/views/sessions/_card.html.erb [hunk 19]
```diff
@@ -1,35 +1,67 @@
...
-    <% end %>
     </div>
-    <div class="flex flex-row items-center mb-1">
-      <%= inline_svg_tag "calendar.svg", class: "fill-grey-600 w-3 h-3 mr-2" %> <p class="text-[13px]"><%= session.starts_at.strftime("%A, %b %d") %></p>
-    </div>
-    <div class="flex flex-row items-center mb-1">
-      <%= inline_svg_tag "clock.svg", class: "stroke-grey-600 fill-grey-600 w-3 h-3 mr-2" %><p class="text-[13px]"><%= session.starts_at.strftime("%H:%M") %> - <%= session.ends_at.strftime("%H:%M") %></p>
+
+    <a href="<%= session_path(session) %>" class="mb-4">
+      <h3 class="text-xl italic font-bold text-red"><%= session&.title %></h3>
```
Comment (Sergio-e) [inline app/views/sessions/_card.html.erb:38]:
```suggestion
      <h3 class="text-xl italic font-bold text-red"><%= session.title %></h3>
```

File: app/views/sessions/_card.html.erb [hunk 20]
```diff
@@ -1,35 +1,67 @@
...
+    </a>
+
+    <div class="flex flex-wrap w-full gap-2 mb-4">
+      <% session.tags&.each do |tag| %>
+        <div class="px-3 py-1.5 text-sm font-medium bg-transparent border rounded-lg border-bluegray-600 text-bluegray-600">
+          <%= tag&.name %>
```
Comment (Sergio-e) [inline app/views/sessions/_card.html.erb:44]:
```suggestion
          <%= tag.name %>
```

File: app/views/sessions/_card.html.erb [hunk 21]
```diff
@@ -1,35 +1,67 @@
...
+          <%= image_tag url_for(speaker&.image), class: "text-red w-12 h-12 rounded-full mr-4" %>
+        <% else %>
+          <div class="flex flex-row items-center justify-center w-12 h-12 p-2 mr-4 rounded-full min-w-12 min-h-12 bg-slate-300">
+            <%= inline_svg_tag "icons/user_avatar.svg" %>
+          </div>
+        <% end %>
+
+        <div class="flex flex-col justify-center h-full">
+          <a href="<%= speaker_path(speaker) %>">
+            <h2 class="italic font-black text-red"><%= speaker&.name %></h2>
```
Comment (Sergio-e) [inline app/views/sessions/_card.html.erb:18]: Seems like the page becomes unusable if at least one session doesn't have a speaker. I think we shouldn't show "incomplete" sessions, that is, those without a location or a speaker. (`@sessions = sessions.joins(:speakers, :location).includes(:attendees, :tags)`
image

File: app/views/sessions/_filters.html.erb [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+
+<div class="flex justify-between w-full mb-6">
+  <div class="flex items-center gap-2 overflow-x-scroll no-scrollbar">
+    <%= link_to(
+      'All dates',
+      url_for(resource),
+      class: [
+        "flex min-w-fit items-center justify-center px-4 py-2 rounded-full font-medium",
+        "bg-red text-white": params[:starts_at].nil?,
```
Comment (Sergio-e) [inline app/views/sessions/_filters.html.erb:10]:
```suggestion
        "bg-red text-white": params[:starts_at].blank?,
```

File: app/views/sessions/index.html.erb [hunk 23]
```diff
@@ -1,3 +1,28 @@
...
+    </div>
+
+    <%= render partial: 'sessions/filters', locals: { resource: :sessions } %>
+
+    <div class="flex flex-col w-full h-full">
+      <% if @sessions.present? %>
+        <%= render partial: 'sessions/card', collection: @sessions, as: :session %>
+      <% elsif params[:starts_at].present? %>
+        <div class="flex flex-col items-center justify-center w-full h-full">
+          <p class="max-w-xs mb-4 font-bold text-center">No sessions found with the selected filters</p>
```
Comment (Sergio-e) [inline app/views/sessions/index.html.erb:14]: Consider using this copy instead "No sessions match your current filters."

File: app/controllers/attendees_controller.rb (hunk 11 above)
Comment (Sergio-e) [reply app/controllers/attendees_controller.rb:27]: I don't think memoization is needed; this isn't an expensive query, nor is it being called multiple times

**Pull request merged**